metals/src/main/scala/scala/meta/internal/metals/debug/DebugServer.scala

```diff
@@ -105,6 +112,155 @@ object DebugServer {
     }
   }
 
+  def resolveMainClassParams(
+      params: DebugUnresolvedMainClassParams,
+      buildTargets: BuildTargets,
+      buildTargetClasses: BuildTargetClasses,
+      showWarningMessage: String => Unit
+  ): Try[b.DebugSessionParams] = {
+    val classAndTarget = Option(params.buildTarget).fold {
+      val classes =
+        buildTargetClasses.findMainClassByName(params.mainClass)
+      if (classes.nonEmpty) Success {
+        val classAndTarget = classes.head
+        if (classes.length > 1) {
+          val className = classAndTarget._1.getClassName()
+          val targetName =
+            buildTargets.info(classAndTarget._2).get.getDisplayName
+          val anotherTargets = classes.tail
+            .map {
+              case (_, id) =>
+                buildTargets.info(id).get.getDisplayName
+            }
+          showWarningMessage(
+            UnresolvedDebugSessionParams
+              .runningClassMultipleBuildTargetsMessage(
+                className,
+                targetName,
+                anotherTargets,
+                "main"
+              )
+          )
+        }
+        classAndTarget
+      }
+      else
+        Failure(new ClassNotFoundException(params.mainClass))
+    } { targetName =>
+      buildTargets
+        .findByDisplayName(targetName)
+        .fold {
+          Failure[(b.ScalaMainClass, b.BuildTargetIdentifier)](
+            new BuildTargetNotFoundException(targetName)
+          ): Try[(b.ScalaMainClass, b.BuildTargetIdentifier)]
+        } { target =>
+          buildTargetClasses
+            .classesOf(target.getId())
+            .mainClasses
+            .values
+            .find(
+              _.getClassName == params.mainClass
+            )
+            .fold {
+              Failure[(b.ScalaMainClass, b.BuildTargetIdentifier)](
+                new ClassNotFoundInBuildTargetException(
+                  params.mainClass,
+                  targetName
+                )
+              ): Try[(b.ScalaMainClass, b.BuildTargetIdentifier)]
+            } { clazz => Success(clazz -> target.getId()) }
+        }
+    }
+    classAndTarget.map {
+      case (clazz, target) =>
+        clazz.setArguments(Option(params.args).getOrElse(List().asJava))
+        clazz.setJvmOptions(Option(params.jvmOptions).getOrElse(List().asJava))
+        val dataKind =
+          b.DebugSessionParamsDataKind.SCALA_MAIN_CLASS
+        val data = clazz.toJson
+        new b.DebugSessionParams(
+          singletonList(target),
+          dataKind,
+          data
+        )
+    }
+  }
```

Comment:
Could you split the method somehow?

Reply:
I've moved finding class and build target data into new class `BuildTargetClassesFinder`.
`findMainClassAndItsBuildTarget` and `findTestClassAndItsBuildTarget` can also probably can be refactored, especially since they follow the same scenario, just main/test is different, but I'm a bit afraid, it eventually get complex-easy (instead of a simple-hard).